sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/ExecutorServiceParallelExecutor.java

```diff
@@ -304,8 +304,10 @@ public void run() {
             visibleUpdates.offer(VisibleExecutorUpdate.fromThrowable(update.getException().get()));
           }
         }
-        fireTimers();
-        mightNeedMoreWork();
+        if (!fireTimers()) {
+          // If any timers have fired, they will add more work; We don't need to add more
+          addWorkIfNecessary();
```

Comment:
Now that I think about this, I'd propose:

```
boolean timersFired = fireTimers();
addWorkIfNecessary(timersFired);
```

This replaces the comment on what `fireTimers()` returns by assigning a value, and it moves all the logic on when to fire timers into `addWorkIfNecessary`.